### PR TITLE
feat: cronjob for streaks

### PR DESCRIPTION
**Pull request by LucasFalkowsky**
closes: https://github.com/corona-school/project-user/issues/1028

What was done:

- cron job list was updated to revalidate streaks by week and month
- manual execution was added to test the functionality
- cron job function was implemented for 2 streak types weekly and monthly
- renaming of awardUser to rewardUser

### CODE REVIEW

File: jobs/list.ts [hunk 1]
```diff
@@ -36,4 +37,7 @@ export const allJobs: CSCronJob[] = [
     { cronTime: '00 00 10 01 * *', jobFunction: postStatisticsToSlack, name: 'postStatisticsToSlack' },
     // Every night, trigger actions for courses that ended yesterday
     { cronTime: '00 00 10 * * *', jobFunction: notificationsEndedYesterday, name: 'notificationsEndedYesterday' },
+    // Revalidate streaks are used to revalidate if streaks are still meeting their conditions and updates them accordingly
+    { cronTime: '00 00 02 * * 1', jobFunction: () => revalidateStreaks(StreakTimeFrame.WEEK), name: 'weeklyStreakRevalidation' },
```
Comment (LomyW) [inline jobs/list.ts:41]: so this mean every monday at 2 a.m. or? Just a question for me to understand the time 😄

File: jobs/list.ts [hunk 2]
```diff
@@ -36,4 +37,7 @@ export const allJobs: CSCronJob[] = [
...
+    { cronTime: '30 01 1 * *', jobFunction: () => revalidateStreaks(StreakTimeFrame.MONTH), name: 'monthlyStreakRevalidation' },
```
Comment (LomyW) [inline jobs/list.ts:42]: same question here: every first day of a month at 1.30 a.m.? 😄

File: jobs/periodic/revalidate-streaks/index.ts [hunk 3]
```diff
@@ -0,0 +1,35 @@
+import { checkUserAchievement } from '../../../common/achievement';
+import { UserAchievementTemplate } from '../../../common/achievement/types';
+import { prisma } from '../../../common/prisma';
+
+export enum StreakTimeFrame {
+    WEEK = 'by_weeks',
+    MONTH = 'by_months',
+}
+
+export default async function revalidateStreaks(timeFrame: StreakTimeFrame) {
```
Comment (LomyW) [inline jobs/periodic/revalidate-streaks/index.ts:10]: can you please add a comment why we need a cronjob for streaks? (if possible 😄 )

File: jobs/periodic/revalidate-streaks/index.ts [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+            context: true,
+            template: true,
+            recordValue: true,
+        },
+    });
+    // This can not be done directly in prisma fetch, because it is not yet possible to filter on the presence of a specific key
+    // https://www.prisma.io/docs/concepts/components/prisma-client/working-with-fields/working-with-json-fields
+    const achievementsToCheck = streakAchievements.filter((achievement) => {
+        const firstKey = Object.keys(achievement.template.conditionDataAggregations)[0];
+        return achievement.template.conditionDataAggregations[firstKey].createBuckets === timeFrame;
```
Comment (LomyW) [inline jobs/periodic/revalidate-streaks/index.ts:30]: I don't understand the `timeFrame`?

**Review by LomyW — CHANGES_REQUESTED** [review]
some questions 🤓

File: jobs/list.ts (hunk 1 above)
Comment (LucasFalkowsky) [reply jobs/list.ts:41]: yes This means every first day of a week ( last number ) at 02 AM

File: jobs/list.ts (hunk 2 above)
Comment (LucasFalkowsky) [reply jobs/list.ts:42]: Also correct! But as you mentioned this. I missed a 00 at the start. The configuration of cronTime works like this:

```
field          allowed values
-----          --------------
second         0-59
minute         0-59
hour           0-23
day of month   1-31
month          1-12 (or names, see below)
day of week    0-7 (0 or 7 is Sunday, or use names)
```

File: jobs/periodic/revalidate-streaks/index.ts (hunk 3 above)
Comment (LucasFalkowsky) [reply jobs/periodic/revalidate-streaks/index.ts:10]: Done!

**Pull request closed without merging**